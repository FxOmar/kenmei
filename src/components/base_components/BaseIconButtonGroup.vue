<template lang="pug">
  span.relative.z-0.inline-flex.shadow-md.rounded-md
    button.group(
      type='button'
      v-for="(button, index) in buttons"
      :key="index"
      :class="buttonClasses(button, index)"
      :content="button.text"
      @click="click(button)"
      v-tippy
    )
      component.h-5.w-5(
        :is="button.icon"
        :class="{ 'opacity-50': button.disabled }"
      )
</template>

<script>
  export default {
    props: {
      buttons: {
        type: Array,
        required: true,
        validator: (prop) => prop.length >= 2,
      },
    },
    methods: {
      click(button) {
        if (button.disabled) return;

        this.$emit('click', button.text);
      },
      buttonClasses(button, index) {
        return {
          'hover_text-gray-800': button.colour === 'primary' && !button.disabled,
          'hover_text-green-400': button.colour === 'success' && !button.disabled,
          'hover_text-red-400': button.colour === 'danger' && !button.disabled,
          'rounded-l-md': index === 0,
          '-ml-px': index > 0,
          'rounded-r-md': index === this.buttons.length - 1,
          disabled: button.disabled,
        };
      },
    },
  };
</script>

<style lang="scss" scoped>
  button {
    @apply relative inline-flex items-center px-2 py-1 text-gray-500;
    @apply border border-gray-300 bg-white text-sm leading-5 font-medium;
    @apply transition ease-in-out duration-150;

    &:focus {
      @apply z-10 outline-none;
    }

    &:active {
      @apply bg-gray-100 text-gray-500;
    }
  }

  .disabled {
    @apply cursor-not-allowed;

    &:active {
      @apply bg-white;
    }
  }
</style>
