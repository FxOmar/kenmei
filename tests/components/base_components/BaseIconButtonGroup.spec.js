import BaseIconButtonGroup from '@/components/base_components/BaseIconButtonGroup.vue';

const localVue = createLocalVue();

localVue.directive('tippy', true);

describe('BaseIconButtonGroup.vue', () => {
  let iconButtonGroup;

  beforeEach(() => {
    iconButtonGroup = shallowMount(BaseIconButtonGroup, {
      localVue,
      propsData: {
        buttons: [
          {
            text: 'Edit',
            icon: 'IconEdit',
            colour: 'primary',
          },
          {
            text: 'Delete',
            icon: 'IconTrash',
            colour: 'danger',
          },
        ],
      },
      stubs: {
        'icon-edit': true,
        'icon-trash': true,
      },
    });
  });

  describe('when buttons provided', () => {
    it('shows icon buttons', () => {
      expect(iconButtonGroup.html()).toMatchSnapshot();
    });

    describe('and a button is clicked', () => {
      it('emits click event with button text', async () => {
        await iconButtonGroup.find('button').trigger('click');

        expect(iconButtonGroup.emitted('click')[0]).toEqual(['Edit']);
      });
    });

    describe('and a button is disabled', () => {
      beforeEach(() => {
        iconButtonGroup.setProps({
          buttons: [
            {
              text: 'Edit',
              icon: 'IconEdit',
              colour: 'primary',
              disabled: true,
            },
          ],
        });
      });

      it('sets it as disabled and disables click event', async () => {
        expect(iconButtonGroup.get('button').element).toHaveClass('disabled');

        await iconButtonGroup.find('button').trigger('click');

        expect(iconButtonGroup.emitted('click')).toBeFalsy();
      });
    });
  });
});
