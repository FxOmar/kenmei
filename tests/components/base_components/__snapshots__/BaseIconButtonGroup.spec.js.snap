// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseIconButtonGroup.vue when buttons provided shows icon buttons 1`] = `<span class="relative z-0 inline-flex shadow-md rounded-md"><button type="button" content="Edit" class="group hover_text-gray-800 rounded-l-md"><iconedit-stub class="h-5 w-5"></iconedit-stub></button><button type="button" content="Delete" class="group hover_text-red-400 -ml-px rounded-r-md"><icontrash-stub class="h-5 w-5"></icontrash-stub></button></span>`;
